### PR TITLE
[RISCV] CSE by swapping conditional branches

### PR DESCRIPTION
**Pull request by mshockwave**
DAGCombiner, as well as InstCombine, tend to canonicalize GE/LE into GT/LT, namely:
```
X >= C --> X > (C - 1)
```
Which sometime generates off-by-one constants that could have been CSE'd with surrounding constants.
Instead of changing such canonicalization, this patch tries to swap those branch conditions post-isel, in the hope of resurfacing more constant CSE opportunities. More specifically, it performs the following optimization:

For two constants C0 and C1 from
```
li Y, C0
li Z, C1
```
To remove redundnat `li Y, C0`,
 1. if C1 = C0 + 1 we can turn: 
    (a) blt Y, X -> bge X, Z
    (b) bge Y, X -> blt X, Z
 2. if C1 = C0 - 1 we can turn: 
    (a) blt X, Y -> bge Z, X
    (b) bge X, Y -> blt Z, X

This optimization will be done by PeepholeOptimizer through RISCVInstrInfo::optimizeCondBranch.

### CODE REVIEW

**Comment by topperc** [comment]
Can you mention the InstCombine/DAGCombine canonicalizations in the description to help motivate this change.

File: llvm/lib/Target/RISCV/RISCVInstrInfo.cpp [hunk 1]
```diff
@@ -1159,6 +1159,133 @@ bool RISCVInstrInfo::reverseBranchCondition(
...
+    MachineInstr *Def = MRI.getVRegDef(Reg);
+    return Def->getOperand(2).getImm();
+  };
+
+  // Try to find the register for constant Z; return
+  // invalid register otherwise.
+  auto searchConst = [&](int64_t C1) -> Register {
+    MachineInstr *DefC1 = nullptr;
+    MachineBasicBlock::reverse_iterator II(&MI), E = MBB->rend();
+    for (++II; II != E; ++II) {
```
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVInstrInfo.cpp:1229]: Can we use std::find_if here?

File: llvm/lib/Target/RISCV/RISCVInstrInfo.cpp [hunk 2]
```diff
@@ -1159,6 +1159,133 @@ bool RISCVInstrInfo::reverseBranchCondition(
...
+      return DefC1->getOperand(0).getReg();
+    else
+      return Register();
+  };
+
+  bool Modify = false;
+  if (isFromLoadImm(LHS) && MRI.hasOneUse(LHS.getReg())) {
+    // Might be case 1.
+    int64_t C0 = getConst(LHS);
+    // Signed integer overflow is UB. (UINT_MAX is bigger so we don't need
```
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVInstrInfo.cpp:1246]: I think you want INT64_MAX?

File: llvm/lib/Target/RISCV/RISCVInstrInfo.cpp (hunk 1 above)
Comment (mshockwave) [reply llvm/lib/Target/RISCV/RISCVInstrInfo.cpp]: Done

File: llvm/lib/Target/RISCV/RISCVInstrInfo.cpp (hunk 2 above)
Comment (mshockwave) [reply llvm/lib/Target/RISCV/RISCVInstrInfo.cpp]: Done

**Comment by mshockwave** [comment]
> Can you mention the InstCombine/DAGCombine canonicalizations in the description to help motivate this change.

It's done now.

File: llvm/lib/Target/RISCV/RISCVInstrInfo.cpp [hunk 3]
```diff
@@ -1159,6 +1159,128 @@ bool RISCVInstrInfo::reverseBranchCondition(
...
+  // To make sure this optimization is really beneficial, we only
+  // optimize for cases where Y had only one use (i.e. only used by the branch).
+
+  // Right now we only care about LI (i.e. ADDI rs, 0)
+  auto isLoadImm = [](const MachineInstr *MI) -> bool {
+    return MI->getOpcode() == RISCV::ADDI && MI->getOperand(1).isReg() &&
+           MI->getOperand(1).getReg() == RISCV::X0;
+  };
+  // Either a load from immediate instruction or X0.
+  auto isFromLoadImm = [&](const MachineOperand &Op) -> bool {
```
[the remark this answers is not in the document]
Comment (mshockwave) [reply llvm/lib/Target/RISCV/RISCVInstrInfo.cpp]: Done

File: llvm/lib/Target/RISCV/RISCVInstrInfo.cpp [hunk 4]
```diff
@@ -1159,6 +1159,126 @@ bool RISCVInstrInfo::reverseBranchCondition(
...
+  // To make sure this optimization is really beneficial, we only
+  // optimize for cases where Y had only one use (i.e. only used by the branch).
+
+  // Right now we only care about LI (i.e. ADDI x0, imm)
+  auto isLoadImm = [](const MachineInstr *MI, int64_t &Imm) -> bool {
+    if (MI->getOpcode() == RISCV::ADDI && MI->getOperand(1).isReg() &&
+        MI->getOperand(1).getReg() == RISCV::X0) {
+      Imm = MI->getOperand(2).getImm();
+      return true;
+    } else {
```
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVInstrInfo.cpp:1202]: No need for else when `if` body returns

File: llvm/lib/Target/RISCV/RISCVInstrInfo.cpp [hunk 5]
```diff
@@ -1159,6 +1159,126 @@ bool RISCVInstrInfo::reverseBranchCondition(
...
+  // invalid register otherwise.
+  auto searchConst = [&](int64_t C1) -> Register {
+    MachineBasicBlock::reverse_iterator II(&MI), E = MBB->rend();
+    auto DefC1 = std::find_if(++II, E, [&](const MachineInstr &I) -> bool {
+      int64_t Imm;
+      return isLoadImm(&I, Imm) && Imm == C1;
+    });
+    if (DefC1 != E)
+      return DefC1->getOperand(0).getReg();
+    else
```
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVInstrInfo.cpp:1232]: You can drop this else

**Review by topperc — APPROVED** [review]
LGTM with those 2 minor else removal comments.

File: llvm/test/CodeGen/RISCV/branch-opt.ll [hunk 6]
```diff
@@ -0,0 +1,175 @@
+; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py UTC_ARGS: --version 3
+; RUN: llc -mtriple=riscv32 -O2 -verify-machineinstrs < %s | FileCheck %s --check-prefix=RV32
+; RUN: llc -mtriple=riscv64 -O2 -verify-machineinstrs < %s | FileCheck %s --check-prefix=RV64
+
+define void @u_case1_a(ptr %a, i32 %b, ptr %c, ptr %d) {
```
Comment (topperc) [inline llvm/test/CodeGen/RISCV/branch-opt.ll:5]: Would adding `signext` attribute to `i32 %b` in all these tests make the output the same for riscv32 and riscv64?

File: llvm/lib/Target/RISCV/RISCVInstrInfo.cpp (hunk 4 above)
Comment (mshockwave) [reply llvm/lib/Target/RISCV/RISCVInstrInfo.cpp:1202]: I believe this is inside `isLoadImm` so we need the else branch to return false.

Comment (mshockwave) [reply llvm/lib/Target/RISCV/RISCVInstrInfo.cpp]: Done

File: llvm/lib/Target/RISCV/RISCVInstrInfo.cpp (hunk 5 above)
Comment (mshockwave) [reply llvm/lib/Target/RISCV/RISCVInstrInfo.cpp]: Done

File: llvm/test/CodeGen/RISCV/branch-opt.ll (hunk 6 above)
Comment (mshockwave) [reply llvm/test/CodeGen/RISCV/branch-opt.ll]: yes they're the same with `signext`. I've merged two FileCheck prefixes into one.

**Review by topperc — APPROVED** [review]
LGTM

**Pull request merged**